core/chains/evm/forwarders/forwarder_manager.go

```diff
@@ -0,0 +1,291 @@
+		}
+		f.ctx, f.cancel = context.WithCancel(context.Background())
+		if len(fwdrs) != 0 {
+			f.initForwardersCache(f.ctx, fwdrs)
+			f.subscribeForwardersLogs(fwdrs)
+		}
+
+		f.authRcvr, err = authorized_receiver.NewAuthorizedReceiver(common.Address{}, f.evmClient)
+		if err != nil {
+			return errors.Wrap(err, "Failed to init AuthorizedReceiver")
+		}
+
+		f.offchainAgg, err = offchain_aggregator_wrapper.NewOffchainAggregator(common.Address{}, f.evmClient)
+		if err != nil {
+			return errors.Wrap(err, "Failed to init OffchainAggregator")
+		}
+
+		f.wg.Add(1)
+		go f.runLoop()
+		return nil
+	})
+}
+
+// TODO(samhassan): this should be aware of job type to decide how to fetch senders list.
+func (f *FwdMgr) MaybeForwardTransaction(from common.Address, to common.Address, encodedPayload []byte) (fwdAddr common.Address, fwdPayload []byte, err error) {
+
+	senders, err := f.getContractSenders(to)
+	if err != nil {
+		return to, encodedPayload, errors.Wrap(err, "Skipping forwarding transaction")
+	}
+
+	// Gets current forwarders that are in `to` senders
+	fwdrs, err := f.ORM.FindForwardersInListByChain(utils.Big(*f.evmClient.ChainID()), senders)
+	if err != nil {
+		return to, encodedPayload, errors.Wrap(err, "Skipping forwarding transaction")
+	}
+
+	for _, fwdr := range fwdrs {
+		eoas, err := f.getContractSenders(fwdr.Address)
+		if err != nil {
+			f.logger.Warnf("Failed to get forwarder senders %s", err)
+			continue
+		}
+		for _, eoa := range eoas {
+			if eoa != from {
+				continue
+			}
+			forwardedPayload, err := f.getForwardedPayload(to, encodedPayload)
+			if err != nil {
+				f.logger.AssumptionViolation("Forwarder encoding failed, this should never happen")
+				continue
+			}
+			return fwdr.Address, forwardedPayload, nil
+		}
+	}
+
+	return to, encodedPayload, errors.New("Skipping forwarding transaction")
+}
+
+func (f *FwdMgr) getForwardedPayload(dest common.Address, origPayload []byte) ([]byte, error) {
+	callArgs, err := forwardABI.Inputs.Pack(dest, origPayload)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to pack forwarder payload")
+	}
+
+	dataBytes := append(forwardABI.ID, callArgs...)
+	return dataBytes, nil
+}
+
+func (f *FwdMgr) getContractSenders(addr common.Address) ([]common.Address, error) {
+	if senders, ok := f.getCachedSenders(addr); ok {
+		return senders, nil
+	}
+	senders, err := f.getAuthorizedSenders(addr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to call getAuthorizedSenders on %s", addr)
+	}
+	f.setCachedSenders(addr, senders)
+	f.subscribeSendersChangedLogs(addr)
+	return senders, nil
+}
+
+func (f *FwdMgr) getAuthorizedSenders(addr common.Address) ([]common.Address, error) {
+	c, err := authorized_receiver.NewAuthorizedReceiverCaller(addr, f.evmClient)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to init forwarder caller")
+	}
+	opts := bind.CallOpts{Context: f.ctx, Pending: false}
```

Comment:
Curious how this deals with re-orgs

Reply:
In general this should follow the logpoller reorgs. I think the way the forwarder is tracking the latest block should be revisited though and I'll look into this after this